feat: Dockerfile!

jamesread opened the pull request:
# What kind of change does this PR introduce?

This is an attempt at the highly requested dockerfile. It's likely to need a few tweaks before merging, but it builds and runs on my machine. 

I did try and copy the raw repo over to Docker and do a `npm install`, but because of the 2Gb of node dependencies (!!!), it actually starts to issue out of file descriptor errors (186k files!). Therefore, it is necessary to do a `npm install` on the host first, then that gets copied over with the build context, and then the nx build completes without issue. 

Things on the todo list;

- [ ] It needs a VOLUME for configuration and persistent data, which is a bit tricky as the .env file is loaded from the same directory as the app at the moment.
- [ ] Split out database migrations, and pause gracefully until migrations are ready. At the moment you'll have to do the migrations sepately until this gets pulled out. If you accept my other PR for the config:check via the commands app, I think it would be good to add a `db:migrate` as a command, or similar. 

# Why was this change needed?

Erm, some people asked for it :-)

nevo-david commented:
This is great :tada:  but a few things:
We need to decide if it's for production or dev.
If it's dev, you don't have to build or copy the directories. All you need to do is run npm run dev.
If it's production, then after the `npm run build`, we need four entry points: 
1. npm run start:prod
2. npm run start:prod:frontend
3. npm run start:prod:workers
4. npm run start:prod:cron

In that case, I think we would have to run docker-compose.

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# Builder image
+FROM foundation AS builder
+
+run mkdir /src
```
RUN ?

jamesread replied:
Good catch, typo!

jamesread replied:
Fixed in b233abaa11b852201bf93b507d35ba088d88718c

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+
+WORKDIR /src
+
+RUN npx nx reset
```
These could be combined

jamesread replied:
It makes sense to combine as much as possible, totally agree, as this keeps the layer count down. 

However, how did you think I should combine these two lines? I'm not seeing it?

jonathan-irvin replied:
This was meant to combine the two run commands.

jamesread replied:
Ah totally, gotcha, will take a look.

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# Foundation image
+FROM registry.fedoraproject.org/fedora-minimal:40 AS foundation
```
Why not just use official node alpine image?

jamesread replied:
I'm personally a fan of Fedora (disclaimer, I also work for Red Hat, but I use Fedora for OliveTin.app and many others), but it's one of the root Linux distributions. The points of parity I see are; 

- Well maintained, committed to free open source licenses, etc
- Minimal image size

The reason I think it accels over arch; 

- It's extensive repositories - I found with OliveTin that you often later get requests for extras, like apprise for notifications, or other bits, and being able to install those just from the package manager really helps. 
- The container registry doesn't require login, like docker.io, so builds wont "randomly" fail after you hit your pull limit and forgot to login! (Minor advantage!)

Do you see Alpine does anything or has any advantages over Fedora that I don't see? I'm open to change if there are technical merits.

scottgrobinson replied:
+1 for using the official node alpine image. Sorry - I think it makes more sense to use the official image here although I understand your reasons for your preference, I feel that's more of a personal preference rather than one that should be in the "official" docker setup.

jamesread replied:
Heya, totally willing to change if there are technical merits to changing. I chose fedora because it technically works really well, so it's become my preference, not because of any feelings. 

This project is likely to require lots of other pieces beyond node in the future, do you see technical reasons why Alpine is better?

jamesread replied:
Switched to arch as per these comments; https://github.com/gitroomhq/postiz-app/pull/205#issuecomment-2336961378

jamesread commented:
> This is great :tada:  but a few things:
> We need to decide if it's for production or dev.
> If it's dev, you don't have to build or copy the directories. All you need to do is run npm run dev.
> If it's production, then after the `npm run build`, we need four entry points: 
> 1. npm run start:prod
> 2. npm run start:prod:frontend
> 3. npm run start:prod:workers
> 4. npm run start:prod:cron
> 
> In that case, I think we would have to run docker-compose.

Heya, many thanks. This is certainly for production/users, as for Dev it's too slow to keep rebuilding the container (the build takes several minutes due to the number of files that need to be copied). 

I was being lazy during testing last night, so I set the ENTRYPOINT to "npm run dev" and forgot to chbsge it before committing. 

You can only have one ENTRYPOINT in a Dockerfile, so I'll find a way to start these 4 in parallel. 

Many many thanks for the comments from @jonathan-irvin too, I see you were interested in this from another discussion but were unsure how to do the npm/nx stuff I believe - that toolchain is also new to me, so this was my best effort! I've tried to address your comments as best I can.

scottgrobinson commented:
> > This is great 🎉  but a few things:
> > We need to decide if it's for production or dev.
> > If it's dev, you don't have to build or copy the directories. All you need to do is run npm run dev.
> > If it's production, then after the `npm run build`, we need four entry points:
> > 
> > 1. npm run start:prod
> > 2. npm run start:prod:frontend
> > 3. npm run start:prod:workers
> > 4. npm run start:prod:cron
> > 
> > In that case, I think we would have to run docker-compose.
> 
> Heya, many thanks. This is certainly for production/users, as for Dev it's too slow to keep rebuilding the container (the build takes several minutes due to the number of files that need to be copied).
> 
> I was being lazy during testing last night, so I set the ENTRYPOINT to "npm run dev" and forgot to chbsge it before committing.
> 
> You can only have one ENTRYPOINT in a Dockerfile, so I'll find a way to start these 4 in parallel.
> 
> Many many thanks for the comments from @jonathan-irvin too, I see you were interested in this from another discussion but were unsure how to do the npm/nx stuff I believe - that toolchain is also new to me, so this was my best effort! I've tried to address your comments as best I can.

I'd suggest docker-compose with 4 seperate containers rather than trying to run them all in one.

jamesread commented:
Hey @scottgrobinson , the image is big, 1.3Gb because of all the dependencies, creating 4 separate containers is of course possible, but the app doesn't yet support scaling in this way really, so what would be gained? It would just use more disk space / RAM.

nevo-david commented:
@jamesread, we should have one for development and one for product.
Production will most likely need four separate containers (because, over time, you will need to scale the workers - microservice).

For dev we don't need 4 commands, we just need to `npm run dev`

jonathan-irvin commented:
I'll take a look at this today and take a stab at a container image.

jamesread commented:
I'm totally open to making changes based on feedback, I just want the best technical decisions to win. 

@nevo-david what's the virtue of having a dev container? It takes a while to build, so it would be frustrating to develop in this way. 

Does the code support active-active (multiple deployments) today? Does your production instance have multiple instances deployed? I ask because there's quite a lot of overhead splitting up the images wheb it comes to running them (the code is the easy bit). There will be people wanting to run this on a raspberry pi for example.

@jonathan-irvin can we work together? I'm on the discord, would love to chat.

nevo-david commented:
@jamesread for the development container, you want to mount the directory but not copy it so people can work on it locally.

jamesread commented:
Heya, just a quick post from me. @nevo-david and I just had a great discussion on discord. 

We discussed several scenarios, and what we roughly said was 3 different Dockerfiles could make sense;

1) dev container, as lots of people are struggling with node/npm versions and installing all the dependencies, this container should bind-mount the apps directory, allowing for easier local dev. 

2) allinone production container (this is what I was trying to do), for small scale self hosters.

3) a scale-out set of containers, which each service separately for bigger scale. 

I don't get to make the final decision, @nevo-david does! I'm also not fussed who makes the commits, I just like being part of helping out in this project :-)

jonathan-irvin commented:
@jamesread I'll handle the scaling side of things with helm and kubernetes. We can accomplish this many ways. Also @nevo-david do we have to host with `nx run frontend:serve:production` or could we serve this with nginx?

jamesread commented:
@jonathan-irvin , massive Kubernetes fan here too, that's where I intend to run it too!

My work here was on the allinone use case, so keen to see what you come up with for the scalable use case if you'd like to do that. Maybe what I started here helps you. 

There is also the dev use case @nevo-david needs as well - maybe we can work together on that?

jonathan-irvin commented:
Yeah, IMO as far as a dev container is concerned, the first thing we can do is make sure the documentation is up-to-date. 

I'm finding some missing dependencies that are assumed. I will put up a PR shortly.

jonathan-irvin commented:
> This is great 🎉 but a few things: We need to decide if it's for production or dev. If it's dev, you don't have to build or copy the directories. All you need to do is run npm run dev. If it's production, then after the `npm run build`, we need four entry points:
> 
> 1. npm run start:prod
> 2. npm run start:prod:frontend
> 3. npm run start:prod:workers
> 4. npm run start:prod:cron
> 
> In that case, I think we would have to run docker-compose.

@nevo-david 

I don't think that docker-compose should be a requirement for building in production. I want to know how to produce artifacts for hosting the frontend, backend, and workers without needing a container to already be running.

jonathan-irvin commented:
I setup #206 for this

jamesread commented:
Okay, it's been a long evening of hacking on [a70ef27](https://github.com/gitroomhq/postiz-app/pull/205/commits/a70ef27cb2a35869e677f18baa021e96c81388dd), but it's 2:45am and I need to go to bed!

There were lots of discussions on discord that led to these changes. Improvements and comments addressed in this commit summarized;

- Base image switched to Arch - @jonathan-irvin made the excellent point that while I know Fedora, fewer people here have experience with it compared to the node:alpine image, and we need something that lots of people know. 
- Rather than the plan of having 3 containers - a devcontainer, allinone and a "scale out", @jonathan-irvin made the excellent suggestion to boot certain services depending on environment variables (defaulting to everything). This means one image that can optionally be scaled service by service, AND run in a allinone configuration as well. This leads to less disk space and wastage, just 2 containers, and a single dockerfile.
- .devcontainer directory added (not really tested much yet), so people  can just `npm install devcontainer; devcontainer up` to start hacking on postiz. 
- Implemented supervisord (still some work to be done there) to start the services
- Switched to `npm run start:prod` etc, rather than using NX, to keep memory usage down, as per @nevo-david 's findings.

TODO;

- I tried for quite a while to inline the node_module dependencies  as part of the `npm run build`, but whatever settings I changed in package.json, the output was always the same. I think this would be good to do, so that we don't have to ship a 2Gb node_modules directory, and just the stuff the build /actually/ depends on.
- Implement supervisord healthchecks. 
- Test both containers like crazy!

jonathan-irvin commented on .devcontainer/devcontainer.json:
I feel like we could engineer this to include everything we need (Postgres and Redis) with some sane defaults sprinkled into the environment variables?

jamesread replied:
So, absolutely. I also haven't started testing the devcontainer much, so at the moment I intended this more like a "just get something there and we'll improve it later".

There are some amazing examples here; https://github.com/JetBrains/devcontainers-examples/tree/main/.devcontainer

We deffo need to add docs as well, but maybe suggest devcontainer improvements come in later PRs, as this PR is already getting quite big?

jonathan-irvin replied:
Yes, let's separate devcontainer into its own thing

jonathan-irvin commented on .dockerignore:
We additionally don't need any markdown files, .devcontainer, .git

jamesread replied:
Great, fixed in 81fd96f6d7f660024ad6f9a0ac14f2c7422f51ef and marked you as co-author.

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# Base image
+FROM docker.io/node:20.17-alpine3.19 AS base
```
nit: add ARG NODE_VERSION=20.17

jamesread replied:
@jonathan-irvin Do you mind just helping me understand what this would give us? It's not possible to change the NODE_VERSION unless we change the base image, right?

jonathan-irvin replied:
It's a dependency and essentially would be used to build multiple / different variants of the container to support future versions of node or test against different versions (like how the LTS version of node 20 had a memory leak and 20.17 worked better).

The kinda flexibility from the get-go really comes in handy later on.

jamesread replied:
Fixed in [7216f21](https://github.com/gitroomhq/postiz-app/pull/205/commits/7216f2135d757046d7cec5373e25294b00389bc0)

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# Base image
+FROM docker.io/node:20.17-alpine3.19 AS base
+
+ENV NPM_CONFIG_UPDATE_NOTIFIER=false
+ENV NEXT_TELEMETRY_DISABLED=1
```
I'm sure these are necessary, can you add comments as to why they are?

jamesread replied:
They're not affecting the build, but it's just turning off unnecessary noise in the build output.

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+EXPOSE 4200
+EXPOSE 3000
+
+RUN mkdir -p /config
```
I think this is redundant with L21 since the `COPY` command should create the directory for you.

jamesread replied:
Sure, fixed in c466f08c80e24d18df5a58ebd0b5409871dd7567

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+COPY var/docker/supervisord /app/supervisord_available_configs/
+COPY .env.example /config/.env
+
+VOLUME /config
```
do we need this as a volume?

jamesread replied:
Yeah, it helps people who use the basic "docker run", or proxmox, or similar from having their config persist. I found quite a few people kind of rely on this feature to have config volumes automatically created for them.

jonathan-irvin commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+
+LABEL org.opencontainers.image.title="Postiz App (DevContainer)"
+
+ENTRYPOINT ["/app/entrypoint.sh"]
```
I think we could consolidate this to the base image since we're extending it on both the devcontainer and dist

jamesread replied:
Yep, great spot. Fixed in 9c9698c59869edeb3df8ebfe949dff810ec4df7f

jonathan-irvin commented on package.json:
```diff
@@ -20,6 +20,8 @@
     "prisma-generate": "cd ./libraries/nestjs-libraries/src/database/prisma && npx prisma generate",
     "prisma-db-push": "cd ./libraries/nestjs-libraries/src/database/prisma && npx prisma db push",
     "prisma-reset": "cd ./libraries/nestjs-libraries/src/database/prisma && npx prisma db push --force-reset && npx prisma db push",
+    "docker-build": "docker rmi localhost/postiz || true && docker build --target dist -t localhost/postiz -f Dockerfile . && docker build --target devcontainer -t localhost/postiz-devcontainer -f Dockerfile .",
+    "docker-create": "docker kill postiz || true && docker rm postiz || true && docker create --name postiz -p 3000:3000 -p 4200:4200 localhost/postiz",
```
nit: consolidate these to shell scripts and have node execute them. Easier to maintain this way.

jonathan-irvin commented on var/docker/entrypoint.sh:
Love this implementation! Can you add some docs on the different setups? Not familiar with supervisor.d but would love to learn!

jonathan-irvin reviewed (changes requested):
Had some time to take a good look at this.

jonathanirvin-ast commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,14 @@
+{
+	"name": "Postiz Dev Container",
+	"image": "localhost/postiz-devcontainer",
```
Adding to this, we can build the Dockerfile and leverage docker-compose in the devcontainer https://containers.dev/implementors/json_reference/#compose-specific

jamesread replied:
This is great find, but can we tweak this in a separate PR focussing on the devcontainer? This PR is already becoming a beast!

jonathan-irvin replied:
Yeah, let's put the devcontainer in a separate PR.

jonathan-irvin commented on .github/workflows/build-containers.yml:
```diff
@@ -0,0 +1,42 @@
+      - name: Setup node
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
```
remember we need to specifically use 20.17

jonathan-irvin commented on .github/workflows/build-containers.yml:
```diff
@@ -0,0 +1,42 @@
+          password: ${{ github.token }}
+
+      - name: docker build
+        run: npm run docker-build
```
I get that we're wrapping docker with npm, but I don't think this is needed.

jonathan-irvin reviewed (changes requested):
A couple more feedback items, but we're looking pretty!

jonathan-irvin commented on .github/workflows/build-containers.yml:
```diff
@@ -0,0 +1,42 @@
+        with:
+          fetch-depth: 0
+
+      - name: Setup node
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20.17'
+          cache: 'npm'
+          cache-dependency-path: |
+            **/package-lock.json
```
We don't need node for building docker

jamesread replied:
Fixed in [79ed890](https://github.com/gitroomhq/postiz-app/pull/205/commits/79ed89060eb7cd819ca527be03799627fc15d71c)

jonathan-irvin reviewed (changes requested):
We don't need node for building docker

jonathan-irvin reviewed (approved):
I had some suggestions to the Dockerfile, but let's not let perfect be the enemy of good. We can iterate on it and make improvements in future PRs.
